### PR TITLE
[bug] fix grad-of-vmap-of-dynamic_slice with out-of-bound indices

### LINKED ISSUE #34228: [bug] grad-of-vmap-of-dynamic_slice with out-of-bound index is incorrect
Example:
```python
In [1]: import jax

In [2]: import jax.numpy as jnp

In [3]: x = jnp.array([1.0])

In [4]: idx = jnp.array([1])

In [5]: def f(x, i):
   ...:   return jax.lax.dynamic_index_in_dim(x, i, axis=0, keepdims=False)
   ...: 

In [6]: jax.jacrev(f)(x, idx[0])  # should be [1.]  ✅
Out[6]: Array([1.], dtype=float32)

In [7]: jax.jacrev(jax.vmap(f, (None, 0)))(x, idx)  # should be [[1.]]  ❌
Out[7]: Array([[0.]], dtype=float32)
```
This is similar to #25878 for `gather`, and it looks like this happens because the `dynamic_slice` batching rule uses `mode=PROMISE_IN_BOUNDS` rather than `mode=CLIP`.

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #34228

The issue stems from the fact that autodiff of `gather` with `mode=PROMISE_IN_BOUNDS` does not behave correctly when the index is out-of-bounds. The semantics of `dynamic_slice` are to clip out-of-bound indices, so we should do this explicitly in the batching rule, so that the gradient of the batching rule reflects the clipping behavior.

### CODE REVIEW

**Review by froystig — APPROVED** [review]

**Comment by hawkinsp** [comment]
> The semantics of dynamic_slice are to clip out-of-bound indices

I don't think this is correct. It looks to me like dynamic_slice has promise_in_bounds semantics?

Maybe it should have clip semantics, but it does not at the moment?

**Comment by jakevdp** [comment]
dynamic_slice has clip semantics outside vmap. This change makes the behavior under vmap consistent with that behavior. There's an example showing this in the attached bug: https://github.com/jax-ml/jax/issues/34228

**Comment by hawkinsp** [comment]
> dynamic_slice has clip semantics outside vmap. This change makes the behavior under vmap consistent with that behavior. There's an example showing this in the attached bug: #34228

As discussed offline, this was not the original intent, but I am ok with changing the intent.
